### PR TITLE
Remove node handler.

### PR DESCRIPTION
**Pull request by alekseysidorov**
Rename to the `TestKit` and avoid of using the `NodeHandler`.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -131,145 +153,232 @@ impl Validator {
...
         );
-        self.validator_count = validator_count;
+        self.validators.extend(additional_validators);
         self
     }
 
-    /// Creates the harness.
-    pub fn create(&self) -> TestHarness {
+    /// Adds a service to the testkit.
+    pub fn with_service(mut self, service: Box<Service>) -> Self {
```
Comment (slowli) [inline src/lib.rs]: Nitpick: maybe, make this method parametric on the service type:

```rust
pub fn with_service<S: Service>(mut self, service: S) -> Self
```

Or use `S: Into<Box<Service>>` (analogous to `Box<Error>` conversions [built into the Rust core](https://doc.rust-lang.org/src/std/error.rs.html#164-168)). Would require changes in the core, though.

File: src/lib.rs [hunk 2]
```diff
@@ -131,145 +153,232 @@ impl Validator {
...
+        self.services.push(service);
+        self
+    }
+
+    /// Creates the testkit.
+    pub fn create(self) -> TestKit {
```
Comment (slowli) [inline src/lib.rs]: ~Nitpick: Is there a particular reason `create()` is consuming? Is just because no one in their right mind would initialize couple of `Testkit`s in a single run? 😄~

**Edit:** It's because `Vec<Box<Service>>` is non-cloneable.

File: src/lib.rs [hunk 3]
```diff
@@ -425,14 +567,31 @@ impl TestHarness {
...
+
+    /// Returns the test node memory pool handle.
+    pub fn mempool(&self) -> RwLockReadGuard<BTreeMap<crypto::Hash, Box<Transaction>>> {
+        self.mempool.read().expect(
+            "Can't read transactions from the mempool.",
+        )
+    }
+
+    /// Returns the leader on the current height. At the moment first validator.
+    pub fn leader(&self) -> &TestNode {
```
Comment (slowli) [inline src/lib.rs:643]: Comment: I don't think having a dedicated `leader()` method is the best idea, if the intent is to customize block creation. A better idea could involve something like a "block builder" which would allow to customize the proposer, round, precommit authors, etc. It can be solved later, though, so for now, `leader()` can stay.

File: src/lib.rs [hunk 4]
```diff
@@ -293,11 +402,24 @@ impl TestHarness {
         // XXX: Creating a new executor each time seems very inefficient, but sharing
         // a single executor seems to be impossible
         // because `handler` needs to be borrowed mutably into the closure. Use `RefCell`?
-        let handler = &mut self.handler;
+        let mempool = Arc::clone(&self.mempool);
+        let snapshot = self.blockchain.snapshot();
+        let schema = CoreSchema::new(snapshot);
         let event_stream = self.channel.api_requests.1.by_ref().greedy_fold(
```
Comment (slowli) [inline src/lib.rs:415]: Note to self: given that `mempool` and `blockchain` are in essence smart pointers, hopefully, it could be possible to avoid instantiation of a `Spawn` on each `poll_events()` call.

File: src/lib.rs [hunk 5]
```diff
@@ -425,14 +567,31 @@ impl TestHarness {
...
         self.do_create_block(&tx_hashes);
     }
+
+    /// Returns the current height of the blockchain.
+    pub fn height(&self) -> Height {
+        self.state().height().next()
+    }
+
+    /// Returns the test node memory pool handle.
+    pub fn mempool(&self) -> RwLockReadGuard<BTreeMap<crypto::Hash, Box<Transaction>>> {
```
Comment (slowli) [inline src/lib.rs:636]: I don't quite like the idea of returning a sync struct here, it seems to give unnecessary details about the testkit implementation. On the positive note, it seems to be safe from deadlocks because of the borrow checker; it seems to be impossible to retain `mempool()`-returned value and at the same time attempt to update the mempool, e.g., via `create_block()`.

I cannot think of a good alternative at the moment. Something like

```rust
pub fn act_on_mempool<T, F>(&self, f: F) -> T
where
    F: FnOnce(&BTreeMap<crypto::Hash, Box<Transaction>>) -> T,
{
    // Create and drop the read guard here
}
```

could probably work, but it seems like a crutch, too.

File: src/lib.rs [hunk 6]
```diff
@@ -131,145 +153,232 @@ impl Validator {
...
     }
 
-    fn with_services<I>(services: I) -> Self
-    where
-        I: IntoIterator<Item = Box<Service>>,
-    {
-        let db = MemoryDB::new();
-        let blockchain = Blockchain::new(Box::new(db), services.into_iter().collect());
-        TestHarnessBuilder::with_blockchain(blockchain)
+    /// Updates the snapshot of the storage.
```
Comment (slowli) [inline src/lib.rs]: Nitpick: this doesn't actually update snapshot, does it?

File: src/lib.rs [hunk 7]
```diff
@@ -370,33 +492,55 @@ impl TestHarness {
...
+                &patch,
+                block_hash,
+                precommits.iter(),
+            )
+            .unwrap();
+        self.do_update();
+    }
+
+    // Update TeskKit state after commit block.
+    fn do_update(&mut self) {
```
Comment (slowli) [inline src/lib.rs]: Nitpick: does `do_update()` need to be a separate method? It seems that it is called from a single place, and it is very short.

**Review by slowli — APPROVED** [review]
Review as of 495a0cf0a80ca3bc0b8b7cbaf919851e826e2dd2. Overall, very nice, except for a couple of nitpicks that could be solved later.

File: src/lib.rs (hunk 6 above)
Comment (alekseysidorov) [reply src/lib.rs]: Yes.

File: src/lib.rs (hunk 1 above)
Comment (alekseysidorov) [reply src/lib.rs]: Second variant does not work because we did not have implementation of `Into<Box<Service>>`. But I can try to fix it in the exonum repo.

Comment (slowli) [reply src/lib.rs]: I don't think it would be worth it for this particular application, to be honest. Now, if the conversion would be used in several other apps, it could be a different talk...

File: src/lib.rs [hunk 8]
```diff
@@ -495,6 +503,35 @@ impl TestKit {
         let height = self.height();
         let last_hash = self.state().last_hash();
 
+        // Try to modify the test network configuration.
```
Comment (slowli) [inline src/lib.rs]: Nitpick: unlike `do_update()`, the code below seems a good candidate for a separate (non-public) method.

File: src/lib.rs [hunk 9]
```diff
@@ -495,6 +503,35 @@ impl TestKit {
...
+        if let Some(cfg_proposal) = self.cfg_proposal.take() {
+            if cfg_proposal.actual_from() == height {
+                // Modify the self configuration
+                self.network.us = cfg_proposal.us;
+                self.network.validators = cfg_proposal.validators;
+                self.service_context.update_node(self.network.us.clone());
+            } else {
+                // Commit configuration proposal
+                let stored = cfg_proposal.stored_configuration().clone();
+                let is_stored_commited = {
```
Comment (slowli) [inline src/lib.rs]: Nitpick: doing this on each height seems inefficient (although, admittedly, it shouldn't matter much for tests). One possible way to deal with this would be to make `self.cfg_proposal` as a 3-variant enum: `Uncommitted(..)`, `Committed(..)` and `None`. Maybe, there is a more obvious solution.

File: src/lib.rs [hunk 10]
```diff
@@ -592,6 +620,149 @@ impl TestKit {
...
+        }
+    }
+
+    /// Returns the testkit node.
+    pub fn us(&self) -> &TestNode {
+        &self.us
+    }
+
+    /// Modifies the testkit node.
+    pub fn set_us(&mut self, us: TestNode) {
```
Comment (slowli) [inline src/lib.rs:711]: I'm having a little issue with `set_us` and `set_validators` as currently coded. Because `TestNode` implements `Clone`, it seems moderately easy to duplicate validators in the config by accident, which doesn't seem to be checked for now.

File: src/lib.rs [hunk 11]
```diff
@@ -592,6 +620,149 @@ impl TestKit {
...
+        let value = serde_json::to_value(config).unwrap();
+        self.stored_configuration.services.insert(id.into(), value);
+    }
+
+    /// Returns the resulting exonum blockchain configuration.
+    pub fn stored_configuration(&self) -> &StoredConfiguration {
+        &self.stored_configuration
+    }
+
+    fn modify_us_role(&mut self) {
```
Comment (slowli) [inline src/lib.rs]: Nitpick: `modify_our_role` sounds better.

File: tests/test_change_configuration.rs [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        let mut cfg = testkit.actual_configuration();
+        let validator = cfg.validators()[1].clone();
+        cfg.set_actual_from(Height(5));
+        cfg.set_validators(vec![validator]);
+        cfg
+    };
+    let stored = proposal.stored_configuration().clone();
+    testkit.propose_configuration_change(proposal);
+
+    for _ in testkit.height().0..6 {
```
Comment (slowli) [inline tests/test_change_configuration.rs]: Nitpick: imo, this calls for a separate method invoked like `testkit.create_blocks_until(Height(6))`; `0..6` part has confused me a little 😄

File: src/lib.rs [hunk 13]
```diff
@@ -592,6 +620,149 @@ impl TestKit {
...
+        for<'de> D: Deserialize<'de>,
+    {
+        let value = self.stored_configuration.services.get(id).expect(
+            "Unable to find configuration for service",
+        );
+        serde_json::from_value(value.clone()).unwrap()
+    }
+
+    /// Modifies the configuration of the service with the given identifier.
+    pub fn set_service_config<D>(&mut self, id: &str, config: D)
```
Comment (slowli) [inline src/lib.rs:774]: Off-topic: I wonder if it would make sense to write

```rust
pub fn set_service_config<S: Service>(&self, config: S::Config) {
    // ...
}
```

if the `Service` trait was declared with an associated `Config` type and [an associated constant](https://blog.rust-lang.org/2017/08/31/Rust-1.20.html) for the service name. Could help to avoid some shots in the leg 🤔

File: src/lib.rs [hunk 14]
```diff
@@ -303,7 +309,7 @@ impl TestKitBuilder {
 
     /// Adds an additional validators.
     pub fn with_validators(mut self, validators_count: u16) -> Self {
-        assert!(validators_count > 1, "At least one validator must be.");
+        assert!(validators_count > 0, "At least one validator must be.");
```
Comment (slowli) [inline src/lib.rs]: Nice catch, missed this in the last review. 👍

**Review by slowli — COMMENTED** [review]
Some new comments as of 500eb8a596dab08768e263bfa0202327d7add3f1.

File: tests/test_change_configuration.rs (hunk 12 above)
Comment (alekseysidorov) [reply tests/test_change_configuration.rs]: What about `testkit.fast_forward_to(Height(6))`?

File: src/lib.rs (hunk 11 above)
Comment (alekseysidorov) [reply src/lib.rs]: `update_our_role` imo is better.

File: src/lib.rs (hunk 10 above)
Comment (alekseysidorov) [reply src/lib.rs:711]: I want to use `Rc<TestNode>` to share node between classes.

File: tests/test_change_configuration.rs (hunk 12 above)
Comment (slowli) [reply tests/test_change_configuration.rs]: Yeah, that would work as well. I'm thinking about a more general method as a possible option, something like

```rust
pub trait TerminationCondition {
    fn terminated(&self, &TestKit) -> bool;
}

// Creates blocks until specified height
impl TerminationCondition for Height { /* ... */ }
// Creates blocks until given configuration becomes current
impl TerminationCondition for StoredConfiguration { /* ... */ }
// Most generic form
impl<F: Fn(&TestKit) -> bool> TerminationCondition for F { /* ... */ }

impl TestKit {
    // ...
    pub fn fast_forward_to<T: TerminationCondition>(condition: T) { /* ... */ }
}
```

Not sure whether such complication would actually be needed, though.

File: src/lib.rs [hunk 15]
```diff
@@ -572,6 +546,33 @@ impl TestKit {
         self.poll_events();
     }
 
+    // Try to modify the test network configuration.
+    fn do_config_update(&mut self) {
```
Comment (slowli) [inline src/lib.rs]: Nitpick: imo, `update_config` sounds better. Also, why "try" in the comment? As far as I see, there is no `Result` returned or possibility to `panic` here. I'd rephrase
> Update test network configuration if such an update has been scheduled with `propose_configuration_change()`.

File: src/lib.rs [hunk 16]
```diff
@@ -572,6 +546,33 @@ impl TestKit {
...
+        let height = self.height();
+        if let Some(cfg_proposal) = self.cfg_proposal.take() {
+            match cfg_proposal {
+                ConfigurationProposalState::Uncommited(cfg_proposal) => {
```
Comment (slowli) [inline src/lib.rs:554]: Nitpick: maybe, `use ConfigurationProposalState::*` within the method?

File: src/lib.rs [hunk 17]
```diff
@@ -572,6 +546,33 @@ impl TestKit {
...
+                    // Commit configuration proposal
+                    let stored = cfg_proposal.stored_configuration().clone();
+                    let mut fork = self.blockchain.fork();
+                    CoreSchema::new(&mut fork).commit_configuration(stored);
+                    let changes = fork.into_patch();
+                    self.blockchain.merge(changes).unwrap();
+                    self.cfg_proposal = Some(ConfigurationProposalState::Commited(cfg_proposal));
+                }
+                ConfigurationProposalState::Commited(ref cfg_proposal)
+                    if cfg_proposal.actual_from() == height => {
```
Comment (slowli) [inline src/lib.rs:564]: This is one ugly formatting (automatic, I suppose) 😕

File: src/lib.rs [hunk 18]
```diff
@@ -679,6 +680,13 @@ pub struct TestNetworkConfiguration {
     stored_configuration: StoredConfiguration,
 }
 
+// A new configuration proposal state
+#[derive(Debug)]
+enum ConfigurationProposalState {
+    Uncommited(TestNetworkConfiguration),
```
Comment (slowli) [inline src/lib.rs]: It's `Committed` and `Uncommitted` (double `t` in both).

File: src/lib.rs [hunk 19]
```diff
@@ -572,13 +572,32 @@ impl TestKit {
...
 
+    /// Creates block with the given transactions.
+    /// Transactions that are in mempool will be ignored.
+    pub fn create_block_with_transactions(&mut self, txs: Vec<Box<Transaction>>) {
+        let tx_hashes = {
+            let mut mempool = self.mempool.write().expect(
+                "Cannot write transactions to mempool",
+            );
+
+            let mut tx_hashes = Vec::new();
```
Comment (slowli) [inline src/lib.rs:584]: Nitpick: maybe, use `Vec::with_capacity(txs.len())` here? We know the size of the `tx_hashes` vector beforehand.

File: src/lib.rs [hunk 20]
```diff
@@ -572,13 +572,32 @@ impl TestKit {
         self.poll_events();
     }
 
+    /// Creates block with the given transactions.
+    /// Transactions that are in mempool will be ignored.
```
Comment (slowli) [inline src/lib.rs]: What does "ignored" mean here? It seems that transactions that are already in the mempool will be added to the block. A more interesting question: what if some of transactions supplied to the method are already committed (i.e., present in one of previous blocks)?

**Review by slowli — COMMENTED** [review]
Some more comments as of 06c13c6c09f4b73f4c89a349697e90eed30ec2b3.

File: src/lib.rs (hunk 18 above)
Comment (alekseysidorov) [reply src/lib.rs]: Fixed

File: src/lib.rs (hunk 17 above)
Comment (alekseysidorov) [reply src/lib.rs:564]: Fixed

File: src/lib.rs (hunk 16 above)
Comment (alekseysidorov) [reply src/lib.rs:554]: Yes.

File: src/lib.rs (hunk 15 above)
Comment (alekseysidorov) [reply src/lib.rs]: Committed

File: src/lib.rs (hunk 19 above)
Comment (alekseysidorov) [reply src/lib.rs:584]: Yes
